### PR TITLE
Advanced sequence multi channel

### LINKED ISSUE #504: Parameter attribute_ids to nidcpower.Session.create_advanced_sequence() should be an enum
The way the C API works, the client is supposed to pass an array of attribute IDs that he/she wants included in the advanced sequence. In our Python API, right now, this is just an int :(

### PR DESCRIPTION
**Pull request by texasaggie97**
- [X] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/nimi-python/blob/master/CONTRIBUTING.md).
- [x] I've updated [CHANGELOG.md](https://github.com/ni/nimi-python/blob/master/CHANGELOG.md) if applicable.
- [X] I've added tests applicable for this pull request

### What does this Pull Request accomplish?
* Makes it easier to create an advanced sequence by allowing the list of properties to be the string names instead of requiring the attribute IDs
* Create step still handled by customer
    * This is to allow multi channel sequences if they are ever supported

### List issues fixed by this Pull Request below, if any.
* Fix #504 

### What testing has been done?
* System
* New example

### CODE REVIEW

File: generated/nidcpower/nidcpower/session.py [hunk 1]
```diff
@@ -4682,6 +4682,36 @@ def export_attribute_configuration_file(self, file_path):
...
+
+        Args:
+            sequence_name (str): Name of sequence.
+
+            property_names (list of str): Sequence.
+
+            set_as_active_sequence (bool): Specifies that this current sequence is active.
+
+        '''
+        # First we need to get all possible properties we might be setting. The way the NI-DCPower C API is designed,
```
Comment (marcoskirsch) [inline generated/nidcpower/nidcpower/session.py]: I think this comment refers to the other discarded implementation from #957

File: generated/nidcpower/nidcpower/session.py [hunk 2]
```diff
@@ -4682,6 +4682,36 @@ def export_attribute_configuration_file(self, file_path):
...
+        # we need to know this upfront in order to call `niDCPower_CreateAdvancedSequence`. In order to find the attribute
+        # ID of each property, we look at the member Attribute objects of Session.
+        attribute_ids_used = set()
+        for prop in property_names:
+            if prop not in Session.__base__.__dict__:
+                raise KeyError('{0} is not an property on the nidcpower.Session'.format(prop))
+            if not isinstance(Session.__base__.__dict__[prop], _attributes.Attribute) and not isinstance(Session.__base__.__dict__[prop], _attributes.AttributeEnum):
+                raise TypeError('{0} is not an attribute type: {1}'.format(prop, type(Session.__base__.__dict__[prop])))
```
Comment (marcoskirsch) [inline generated/nidcpower/nidcpower/session.py]: "attribute" as a term should never used in customer facing Python code.
"is not a valid property" would work imo

File: generated/nidcpower/nidcpower/session.py [hunk 3]
```diff
@@ -4682,6 +4682,36 @@ def export_attribute_configuration_file(self, file_path):
...
+            attribute_ids_used.add(Session.__base__.__dict__[prop]._attribute_id)
+
+        # Create the sequence with the list of attr ids we have
```
Comment (marcoskirsch) [inline generated/nidcpower/nidcpower/session.py]: comment seems unecessary

File: generated/nidcpower/nidcpower/session.py [hunk 4]
```diff
@@ -4682,6 +4682,36 @@ def export_attribute_configuration_file(self, file_path):
...
+        self._create_advanced_sequence(sequence_name, list(attribute_ids_used), set_as_active_sequence)
```
Comment (marcoskirsch) [inline generated/nidcpower/nidcpower/session.py]: I don't see removing the default parameter from `_create_advanced_sequence`, we should so there's less metadata overall.

File: src/nidcpower/examples/nidcpower_advanced_sequence.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
+#!/usr/bin/python
+
+import argparse
+import datetime
+import nidcpower
+import sys
+
+
+def example(resource_name, channels, options, voltage_max, current_max, points_per_output_function, delay):
+    timeout = datetime.timedelta(seconds=(delay + 1.0))
```
Comment (marcoskirsch) [inline src/nidcpower/examples/nidcpower_advanced_sequence.py]: Why `+1`?

Comment (marcoskirsch) [reply src/nidcpower/examples/nidcpower_advanced_sequence.py]: Why is `delay` stored in `timeout`? They are different concepts.

File: src/nidcpower/examples/nidcpower_advanced_sequence.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
+#!/usr/bin/python
+
+import argparse
+import datetime
+import nidcpower
+import sys
+
+
+def example(resource_name, channels, options, voltage_max, current_max, points_per_output_function, delay):
```
Comment (marcoskirsch) [inline src/nidcpower/examples/nidcpower_advanced_sequence.py]: delay_in_seconds ?
unless client can pass a timedelta.

File: src/nidcpower/examples/nidcpower_advanced_sequence.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        line_format = '{:,<4} {:,.6g} {:,.6g} {:<6}\n'
+        print('{:<4} {:<10} {:,<10} {:<6}'.format('Num', 'Voltage', 'Current', 'In Compliance'))
+        i = 0
+        for measurement in measurements:
+            print(line_format.format(i, measurement.voltage, measurement.current, str(measurement.in_compliance)))
+            i += 1
+
+
+def _main(argsv):
+    parser = argparse.ArgumentParser(description='Outputs voltage 1, waits for source delay, and then takes a measurement. Then orepeat with voltage 2.', formatter_class=argparse.ArgumentDefaultsHelpFormatter)
```
Comment (marcoskirsch) [inline src/nidcpower/examples/nidcpower_advanced_sequence.py]: I don't think this is accurate... it appears to output a ramp.

File: src/nidcpower/system_tests/test_system_nidcpower.py [hunk 8]
```diff
@@ -267,12 +267,13 @@ def test_import_export_file(single_channel_session):
...
+    properties_used = ['output_function', 'voltage_level']
     single_channel_session.source_mode = nidcpower.SourceMode.SEQUENCE
-    single_channel_session._create_advanced_sequence(sequence_name='my_sequence', attribute_ids=ids, set_as_active_sequence=True)
-    single_channel_session._create_advanced_sequence_step(set_as_active_step=True)
+    single_channel_session.create_advanced_sequence(sequence_name='my_sequence', property_names=properties_used, set_as_active_sequence=True)
+    single_channel_session.create_advanced_sequence_step(set_as_active_step=True)
+    single_channel_session.output_function = nidcpower.OutputFunction.DC_VOLTAGE
     single_channel_session.voltage_level = 1
-    single_channel_session._delete_advanced_sequence(sequence_name='my_sequence')
+    single_channel_session.delete_advanced_sequence(sequence_name='my_sequence')
```
Comment (marcoskirsch) [inline src/nidcpower/system_tests/test_system_nidcpower.py]: after creating, read back the active sequence and assert it's correct.
after deleting, set active sequence to the now deleted one and assert an error is raised.

Do we have coverage for things like mispelling an attribute name?

File: docs/nidcpower/class.rst [hunk 9]
```diff
@@ -274,6 +274,142 @@ configure_aperture_time
...
+    .. py:currentmodule:: nidcpower.Session
+
+    .. py:method:: create_advanced_sequence(sequence_name, property_names, set_as_active_sequence=True)
+
+            Test.
+
+            
+
+            .. note:: Test.
+
```
[the remark this answers is not in the document]
Comment (marcoskirsch) [reply docs/nidcpower/class.rst]: ok

File: generated/nidcpower/nidcpower/session.py (hunk 4 above)
Comment (marcoskirsch) [reply generated/nidcpower/nidcpower/session.py]: Let's remove the default parameter from `_create_advanced_sequence`. Having default parameters requires extra metadata, and we don't need it for this private method.

Comment (texasaggie97) [reply generated/nidcpower/nidcpower/session.py]: Removed

File: docs/nidcpower/class.rst (hunk 9 above)
[the remark this answers is not in the document]
Comment (texasaggie97) [reply docs/nidcpower/class.rst]: Documentation updated

File: CHANGELOG.md [hunk 10]
```diff
@@ -47,6 +47,9 @@ All notable changes to this project will be documented in this file.
     * #### Removed
 * ### NI-DCPower
     * #### Added
+        * `create_advanced_sequence()` - [#504](https://github.com/ni/nimi-python/issues/504)
```
Comment (marcoskirsch) [inline CHANGELOG.md]: you also added a couple of attributes, delete_advanced_sequence...

File: src/nidcpower/examples/nidcpower_advanced_sequence.py [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+        for measurement in measurements:
+            print(line_format.format(i, measurement.voltage, measurement.current, str(measurement.in_compliance)))
+            i += 1
+
+
+def _main(argsv):
+    parser = argparse.ArgumentParser(description='Output ramping voltage to voltage max, then ramping current to current max.', formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+    parser.add_argument('-n', '--resource-name', default='PXI1Slot2', help='Resource name of a National Instruments SMU')
+    parser.add_argument('-c', '--channels', default='0', help='Channel(s) to use')
+    parser.add_argument('-p', '--number-points', default=256, help='Number of points per output function')
```
Comment (marcoskirsch) [inline src/nidcpower/examples/nidcpower_advanced_sequence.py]: points should be steps

File: CHANGELOG.md (hunk 10 above)
Comment (texasaggie97) [reply CHANGELOG.md]: Added

**Pull request merged**